### PR TITLE
Write out plots of test set performance and push in `skynet`

### LINKED ISSUE #1225: Add ability to dump test, train and val sets to `.csv` files as part of ML model training
We currently use Torch `Datasets` to iterate through our training data (as we should), but it would be nice to have a convenience method to dump the splits to CSV's that we can then re-use down the track. 

Either myself or @kaminow to tackle.

### PR DESCRIPTION
**Pull request by hmacdope**
Fixes #1225

## Developers certificate of origin
- [X] I certify that this contribution is covered by the MIT license as defined in our [LICENSE](https://github.com/choderalab/asapdiscovery/blob/main/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

### CODE REVIEW

**Comment by hmacdope** [comment]
@kaminow could I grab a review here?

File: asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py [hunk 1]
```diff
@@ -642,20 +828,56 @@ def train_GAT_for_endpoint(
...
+        s3_plot_ens_dest = f"{protocol}/{model_tag}/test_performance.png"
+        logger.info(f"Pushing test performance plot to S3 at {s3_plot_ens_dest}")
+        s3.push_file(plot_path, s3_plot_ens_dest)
+
+    # log the artifact to WandB run
+    logger.info("Logging test performance plot to WandB")
+    for run_id in wandb_run_ids:
+        if run_id is None:
+            continue
+        wandb.init(id=run_id, project=wandb_project, resume="must")
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py:881]: this can raise an error I think if the W&B specification doesn't match an existing run (I think `wandb.errors.UsageError`), can you add a quick `try...except` block to handle that?

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 2]
```diff
@@ -215,6 +215,40 @@ def pull(self, local_dir: Union[Path, str] = None) -> "LocalEnsembleMLModelSpec"
             **self.dict(exclude={"models"}),
         )
 
+    def pull_plot(
+        self, plotname: str, filename: Optional[str] = None, return_as="memory"
+    ):
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/models.py:220]: docstring pls

File: asapdiscovery-ml/asapdiscovery/ml/models.py [hunk 3]
```diff
@@ -215,6 +215,40 @@ def pull(self, local_dir: Union[Path, str] = None) -> "LocalEnsembleMLModelSpec"
...
+            # save to file
+            if not filename:
+                filename = plotname
+            with open(filename, "wb") as f:
+                f.write(response.content)
+        elif return_as == "url":
+            return plot_url
+        else:
+            raise ValueError("return_as must be 'memory' or 'file' or 'url'")
+        return plotname
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/models.py:250]: should this be `filename`?

Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/models.py:250]: also, could be moved up into the relevant `elif` block since all the other code paths have their own return statements anyway

**Review by kaminow — CHANGES_REQUESTED** [review]
looks good, just some relatively minor things

File: asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py [hunk 4]
```diff
@@ -46,6 +53,163 @@
...
+        ax.errorbar(
+            df[readout_column],
+            df[pred_column],
+            xerr=df[xerr_column],
+            fmt="none",
+            capsize=5,
+            zorder=1,
+            color="C0",
+        )
+    stats_dict = do_stats(df[readout_column], df[pred_column])
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py:121]: We can simplify by unifying with your and my code once #1081 is merged.

File: asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py [hunk 5]
```diff
@@ -46,6 +53,163 @@
 SKYNET_SERVE_URL = "https://asap-discovery-ml-skynet.asapdata.org"
 
 
+def plot_test_performance(test_csv, readout_column, model, output_dir):
+    df = pd.read_csv(test_csv)
+    inference_cls = GATInference.from_ml_model_spec(model)
+    smiles = df["smiles"]
+    pred, err = inference_cls.predict_from_smiles(smiles, return_err=True)
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py:60]: Split them

**Comment by hmacdope** [comment]
should be good for another look @kaminow :)

File: asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py:63]: fab ty

File: asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py [hunk 6]
```diff
@@ -46,6 +54,191 @@
 SKYNET_SERVE_URL = "https://asap-discovery-ml-skynet.asapdata.org"
 
 
+def evaluate_test_performance(
+    test_csv, readout_column, model, output_dir, exp_err_column=None
+):
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py:59]: docstring pls

File: asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py [hunk 7]
```diff
@@ -46,6 +54,191 @@
...
+def plot_test_performance(
+    df,
+    model_name,
+    readout_column,
+    pred_column,
+    output_dir,
+    err_column,
+    exp_err_column=None,
+    plotname="test_performance.png",
+) -> str:
```
Comment (kaminow) [inline asapdiscovery-ml/asapdiscovery/ml/cli_mlops.py:89]: docstring pls

**Review by kaminow — APPROVED** [review]
looks good! just a couple missing docstrings but other than that I think good to go
